File: 21.merge-two-sorted-lists/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
	ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
		if (l1 == NULL)
			return l2;
		if (l2 == NULL)
			return l1;
		ListNode *head = l1->val <= l2->val ? l1 : l2;
		ListNode *p = head;
		while (l1 != NULL || l2 != NULL)
		{
			ListNode *tmp;
			if (l1 == NULL || (l2 != NULL && l1->val > l2->val))
			{
				tmp = l2;
				l2 = l2->next;
			} else
			{
				tmp = l1;
				l1 = l1->next;
			}
			p->next = tmp;
			p = p->next;
		}
		return head;
	}
};

File: 21.merge-two-sorted-lists/solution.kt
/**
 * Example:
 * var li = ListNode(5)
 * var v = li.`val`
 * Definition for singly-linked list.
 * class ListNode(var `val`: Int) {
 *     var next: ListNode? = null
 * }
 */
class Solution {
    fun mergeTwoLists(list1: ListNode?, list2: ListNode?): ListNode? {
        val root = ListNode(0)
        var p = root
        var p1 = list1
        var p2 = list2
        while (p1 != null || p2 != null) {
            if (p1 == null || p2 != null && p1.`val` > p2.`val`) {
                p.next = p2
                p2 = p2!!.next
            } else {
                p.next = p1
                p1 = p1.next
            }
            p.next!!.next = null
            p = p.next!!
        }
        return root.next
    }
}
